File: frontend/js/views/three_fund_portfolio.js
// Three-fund portfolio recommender (Boglehead). US + International + Bonds.

import { api } from '../api.js';
import { esc } from '../util.js';
import { t } from '../i18n.js';

export async function renderThreeFundPortfolio(mount, _state) {
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.three_fund_portfolio.title">// THREE-FUND PORTFOLIO</span></h1>
        <p class="muted small" data-i18n-html="view.three_fund_portfolio.intro">
            Canonical Boglehead three-fund portfolio: US stocks + International stocks +
            Bonds. Stock allocation by age + risk tolerance: <strong>aggressive</strong>
            110 − age, <strong>moderate</strong> 100 − age, <strong>conservative</strong>
            90 − age (clamped [10, 95]%). Within stocks default 70/30 US/international
            (Bogle); Vanguard TDF uses 60/40. Reports drift from target + suggested
            rebalance trades.
        </p>
        <div class="chart-panel">
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:8px;margin-bottom:12px">
                <label><span class="muted small" data-i18n="view.three_fund_portfolio.field.age">Age</span>
                    <input type="number" id="tf-age" step="1" min="1" max="110" value="40" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.three_fund_portfolio.field.risk">Risk tolerance</span>
                    <select id="tf-risk">
                        <option value="conservative">Conservative (90 − age)</option>
                        <option value="moderate" selected>Moderate (100 − age)</option>
                        <option value="aggressive">Aggressive (110 − age)</option>
                    </select>
                </label>
                <label><span class="muted small" data-i18n="view.three_fund_portfolio.field.us_share">US share of stocks %</span>
                    <input type="number" id="tf-us-share" step="5" min="0" max="100" value="70" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.three_fund_portfolio.field.us">Current US stocks $</span>
                    <input type="number" id="tf-us" step="1000" min="0" value="100000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.three_fund_portfolio.field.intl">Current Intl stocks $</span>
                    <input type="number" id="tf-intl" step="1000" min="0" value="20000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.three_fund_portfolio.field.bonds">Current Bonds $</span>
                    <input type="number" id="tf-bonds" step="1000" min="0" value="30000" style="width:100%"></label>
            </div>
            <button class="btn btn-sm primary" id="tf-run" data-shortcut="r" data-i18n="view.three_fund_portfolio.btn.run">⚡ Compute Allocation</button>
            <div id="tf-result"></div>
        </div>
    `;
    mount.querySelector('#tf-run').addEventListener('click', () => runCompute(mount));
    await runCompute(mount);
}

async function runCompute(mount) {
    const result = mount.querySelector('#tf-result');
    const input = {
        age: parseInt(mount.querySelector('#tf-age').value, 10) || 0,
        risk_tolerance: mount.querySelector('#tf-risk').value,
        current_us_stocks_usd: parseFloat(mount.querySelector('#tf-us').value) || 0,
        current_intl_stocks_usd: parseFloat(mount.querySelector('#tf-intl').value) || 0,
        current_bonds_usd: parseFloat(mount.querySelector('#tf-bonds').value) || 0,
        us_within_stocks_pct: parseFloat(mount.querySelector('#tf-us-share').value) || 70,
    };
    result.innerHTML = `<p class="muted">${esc(t('view.three_fund_portfolio.status.computing'))}</p>`;
    try {
        const r = await api.request('/three-fund-portfolio/compute', { method: 'POST', body: JSON.stringify(input) });
        const assetRow = (label, a) => `
            <tr>
                <td><strong>${esc(label)}</strong></td>
                <td>${a.target_weight_pct.toFixed(1)}%</td>
                <td>$${(a.target_dollar_usd / 1000).toFixed(1)}K</td>
                <td>$${(a.current_dollar_usd / 1000).toFixed(1)}K</td>
                <td>${a.current_weight_pct.toFixed(1)}%</td>
                <td class="${Math.abs(a.drift_pct) > 5 ? 'neg' : ''}">${a.drift_pct >= 0 ? '+' : ''}${a.drift_pct.toFixed(1)}%</td>
                <td class="${a.rebalance_buy_sell_usd >= 0 ? 'pos' : 'neg'}">${a.rebalance_buy_sell_usd >= 0 ? 'BUY $' : 'SELL $'}${(Math.abs(a.rebalance_buy_sell_usd) / 1000).toFixed(1)}K</td>
            </tr>
        `;
        result.innerHTML = `
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:12px;margin-top:1rem">
                <div><div class="muted small">${esc(t('view.three_fund_portfolio.field.total'))}</div>
                    <strong style="font-size:1.4em">$${(r.total_portfolio_usd / 1000).toFixed(1)}K</strong></div>
                <div><div class="muted small">${esc(t('view.three_fund_portfolio.field.stock_target'))}</div>
                    <strong>${r.total_stock_target_pct.toFixed(0)}%</strong></div>
                <div><div class="muted small">${esc(t('view.three_fund_portfolio.field.bond_target'))}</div>
                    <strong>${r.total_bond_target_pct.toFixed(0)}%</strong></div>
            </div>
            <h2 style="margin-top:1rem">${esc(t('view.three_fund_portfolio.h2.allocation'))}</h2>
            <table class="trades">
                <thead><tr>
                    <th data-i18n="view.three_fund_portfolio.th.asset">Asset</th>
                    <th data-i18n="view.three_fund_portfolio.th.target_pct">Target %</th>
                    <th data-i18n="view.three_fund_portfolio.th.target_dollar">Target $</th>
                    <th data-i18n="view.three_fund_portfolio.th.current_dollar">Current $</th>
                    <th data-i18n="view.three_fund_portfolio.th.current_pct">Current %</th>
                    <th data-i18n="view.three_fund_portfolio.th.drift">Drift</th>
                    <th data-i18n="view.three_fund_portfolio.th.rebalance">Rebalance</th>
                </tr></thead>
                <tbody>
                    ${assetRow(t('view.three_fund_portfolio.asset.us'), r.us_stocks)}
                    ${assetRow(t('view.three_fund_portfolio.asset.intl'), r.intl_stocks)}
                    ${assetRow(t('view.three_fund_portfolio.asset.bonds'), r.bonds)}
                </tbody>
            </table>
        `;
    } catch (e) {
        result.innerHTML = `<p class="neg">${esc(String(e))}</p>`;
    }
}

File: crates/traderview-db/src/three_fund_portfolio.rs
//! Three-fund portfolio recommender (Boglehead).
//!
//! The canonical Boglehead three-fund portfolio: US stocks +
//! International stocks + Bonds. Allocation depends on age + risk
//! tolerance. Heuristics:
//!
//!   - Total stock allocation = 110 − age (aggressive), 100 − age
//!     (moderate), or 90 − age (conservative). Clamped to [10, 95].
//!   - Within stocks, 70/30 US/international (Bogle's recommendation;
//!     Vanguard target-date funds use 60/40).
//!   - Bonds = 100 − stocks.
//!
//! Compute returns recommended weights, target dollar amounts given
//! a portfolio total, current weights from user holdings, drift
//! (current − target), and suggested rebalance trades.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct ThreeFundInput {
    pub age: u32,
    pub risk_tolerance: String, // "conservative" | "moderate" | "aggressive"
    pub current_us_stocks_usd: f64,
    pub current_intl_stocks_usd: f64,
    pub current_bonds_usd: f64,
    /// Within stocks, fraction allocated to US (rest to international).
    /// 0.70 = 70/30 US/intl (Bogle); 0.60 = Vanguard TDF default.
    #[serde(default = "default_us_share")]
    pub us_within_stocks_pct: f64,
}

fn default_us_share() -> f64 { 70.0 }

#[derive(Debug, Clone, Serialize)]
pub struct AssetReport {
    pub target_weight_pct: f64,
    pub target_dollar_usd: f64,
    pub current_dollar_usd: f64,
    pub current_weight_pct: f64,
    pub drift_pct: f64,
    pub rebalance_buy_sell_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct ThreeFundReport {
    pub total_portfolio_usd: f64,
    pub us_stocks: AssetReport,
    pub intl_stocks: AssetReport,
    pub bonds: AssetReport,
    pub total_stock_target_pct: f64,
    pub total_bond_target_pct: f64,
}

// ─── Pure compute ─────────────────────────────────────────────────────────

pub fn total_stock_pct(age: u32, risk: &str) -> f64 {
    let raw = match risk {
        "aggressive" => 110.0_f64 - age as f64,
        "conservative" => 90.0 - age as f64,
        _ => 100.0 - age as f64, // moderate default
    };
    raw.clamp(10.0, 95.0)
}

pub fn compute(input: &ThreeFundInput) -> ThreeFundReport {
    let total = input.current_us_stocks_usd + input.current_intl_stocks_usd
        + input.current_bonds_usd;
    let stock_pct = total_stock_pct(input.age, &input.risk_tolerance);
    let bond_pct = 100.0 - stock_pct;
    let us_share = input.us_within_stocks_pct.clamp(0.0, 100.0);
    let us_target_pct = stock_pct * us_share / 100.0;
    let intl_target_pct = stock_pct * (100.0 - us_share) / 100.0;

    let mk = |target_pct: f64, current: f64| -> AssetReport {
        let target_dollar = total * target_pct / 100.0;
        let current_pct = if total > 0.0 { current / total * 100.0 } else { 0.0 };
        let drift = current_pct - target_pct;
        let buy_sell = target_dollar - current;
        AssetReport {
            target_weight_pct: target_pct,
            target_dollar_usd: target_dollar,
            current_dollar_usd: current,
            current_weight_pct: current_pct,
            drift_pct: drift,
            rebalance_buy_sell_usd: buy_sell,
        }
    };
    ThreeFundReport {
        total_portfolio_usd: total,
        us_stocks: mk(us_target_pct, input.current_us_stocks_usd),
        intl_stocks: mk(intl_target_pct, input.current_intl_stocks_usd),
        bonds: mk(bond_pct, input.current_bonds_usd),
        total_stock_target_pct: stock_pct,
        total_bond_target_pct: bond_pct,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn input() -> ThreeFundInput {
        ThreeFundInput {
            age: 40,
            risk_tolerance: "moderate".into(),
            current_us_stocks_usd: 100_000.0,
            current_intl_stocks_usd: 20_000.0,
            current_bonds_usd: 30_000.0,
            us_within_stocks_pct: 70.0,
        }
    }

    #[test]
    fn total_stock_pct_moderate_basic() {
        // age 40, moderate = 100 − 40 = 60%
        assert_eq!(total_stock_pct(40, "moderate"), 60.0);
    }

    #[test]
    fn total_stock_pct_aggressive_basic() {
        // age 40, aggressive = 110 − 40 = 70%
        assert_eq!(total_stock_pct(40, "aggressive"), 70.0);
    }

    #[test]
    fn total_stock_pct_conservative_basic() {
        // age 40, conservative = 90 − 40 = 50%
        assert_eq!(total_stock_pct(40, "conservative"), 50.0);
    }

    #[test]
    fn total_stock_pct_clamps_high() {
        // age 5 aggressive = 105 → clamped to 95
        assert_eq!(total_stock_pct(5, "aggressive"), 95.0);
    }

    #[test]
    fn total_stock_pct_clamps_low() {
        // age 90 conservative = 0 → clamped to 10
        assert_eq!(total_stock_pct(90, "conservative"), 10.0);
    }

    #[test]
    fn total_stock_pct_unknown_risk_defaults_moderate() {
        assert_eq!(total_stock_pct(40, "bogus"), 60.0);
    }

    #[test]
    fn compute_total_basic() {
        let r = compute(&input());
        assert_eq!(r.total_portfolio_usd, 150_000.0);
    }

    #[test]
    fn compute_target_weights_sum_to_100() {
        let r = compute(&input());
        let sum = r.us_stocks.target_weight_pct + r.intl_stocks.target_weight_pct
            + r.bonds.target_weight_pct;
        assert!((sum - 100.0).abs() < 1e-6);
    }

    #[test]
    fn compute_us_target_70pct_of_stocks() {
        let r = compute(&input());
        // 60% stocks × 70% US within = 42%
        assert!((r.us_stocks.target_weight_pct - 42.0).abs() < 1e-6);
        // 60% stocks × 30% intl within = 18%
        assert!((r.intl_stocks.target_weight_pct - 18.0).abs() < 1e-6);
        // 100 − 60 = 40% bonds
        assert!((r.bonds.target_weight_pct - 40.0).abs() < 1e-6);
    }

    #[test]
    fn compute_drift_basic() {
        let r = compute(&input());
        // Current US: 100k / 150k = 66.67%, target 42% → drift +24.67%
        let curr = 100_000.0 / 150_000.0 * 100.0;
        assert!((r.us_stocks.drift_pct - (curr - 42.0)).abs() < 1e-6);
    }

    #[test]
    fn compute_rebalance_buy_sell_signs() {
        let r = compute(&input());
        // US over-allocated → sell (negative buy_sell)
        assert!(r.us_stocks.rebalance_buy_sell_usd < 0.0);
        // Intl: 20k current, target 18% of 150k = 27k → buy 7k
        assert!(r.intl_stocks.rebalance_buy_sell_usd > 0.0);
        assert!((r.intl_stocks.rebalance_buy_sell_usd - 7_000.0).abs() < 1.0);
        // Bonds: 30k current, target 40% of 150k = 60k → buy 30k
        assert!((r.bonds.rebalance_buy_sell_usd - 30_000.0).abs() < 1.0);
    }

    #[test]
    fn compute_zero_portfolio_safe() {
        let r = compute(&ThreeFundInput {
            age: 40,
            risk_tolerance: "moderate".into(),
            current_us_stocks_usd: 0.0,
            current_intl_stocks_usd: 0.0,
            current_bonds_usd: 0.0,
            us_within_stocks_pct: 70.0,
        });
        assert_eq!(r.total_portfolio_usd, 0.0);
        assert_eq!(r.us_stocks.current_weight_pct, 0.0);
    }

    #[test]
    fn compute_old_age_high_bond_allocation() {
        let mut i = input();
        i.age = 75;
        // moderate, age 75 = 100 − 75 = 25% stocks → 75% bonds
        let r = compute(&i);
        assert!((r.total_bond_target_pct - 75.0).abs() < 1e-6);
    }
}
